UI: Revision of Mainbar

nhaagen opened the pull request:
This is a first (working) example on how states for the mainbar-entries might be stored in a cookie.
It feels like there is still quite a bit room for improvement, but I could use some comments if this is going in the desired direction at all (@chfsx ?).

Thanks a bunch,
Nils

- enumerate entries and sub-slates
- internal data (JS) to group triggerer, slate and close-button
- revise JS to not operate directly on DOM but do calculations first.
- use ilUtilities from GS to store cookie
- ... still to be improved

klees commented:
Thx. I will have a look today!

Amstutz commented:
This has still the "Draft". Something required from my part here?

klees commented:
Hi JF,

we would like to add a new public method [`getToggleToolsSignal`](https://github.com/ILIAS-eLearning/ILIAS/pull/2312/files#diff-0386d22af43419b240ceabba9b0bf780R120) to the mainbar.

Regards!

Amstutz commented:
Hi @Nils

Thanks a lot for this!

Note the getToggleToolsSignal is an interface change and needs JF blessing. Therefore I put it on the JF agenda. @klees, fine with you? Do we need some ruling to let minor interface changes pass without JF blessing, so we do not block dev. for 14 days each time we have a slight change of the interface and simply announce it afterwards? However, then we fall into the pit of having to decide each time of what is minor and major. I am not sure about it.

Here some questions:
 - [x] ID’s: You introduce IDs on the immutable objects. I see the reason, they are def. needed. However, Will the be reusable nevertheless? Why did you not use a name source like pattern, as we did with the Inputs?

Some proposals, plz give reason if ignored:
 - [x] Classes in PHP: I dislike parts such as: $('#{$id}').addClass('{$init_state}‘);. I rather prefer a public interface of your JS code with function explicitely telling, what is done. Like that, you also will not have to use classes defined in PHP. E.g. here you could offer a function such as initSlateButton(state,mb_id,mb_parent,signal) or similar. Same bellow, IMO we should try to keep the OnLoad Codes one liners piping arguments through to Functions defined on the JS side, keep the interface small and clear.

And finally the change requests:
 - [x] ID’s: You introduce 2 ID’s into the MainBars Component implementations. What are those? DOM ID’s? Global Screen ID’s? Please explain in the PHP-Doc section properly. Plz also describe why they are needed.
 - [x] ID’s: Same story with the Ids on the slate
 - [x] JS: To be honest, I do not understand a lot of what is going on in the JS-Code. As far as I can see a big jQuery blood baths. However, who am I to tell, I did something very similar, if not worse just hours ago: https://github.com/ILIAS-eLearning/ILIAS/pull/2329. IMO, sadly to only way with our patterns at hand to get some tasks done. However, what would help tremendously, if you could document for each method on the public interface a) why it is publicly exposed to the outside (e.g. entries are a rather broad exposure), b) what they can be used for and c) what they do. Namely, do that for the following: entries: entries, registerEntry, render, ddTriggerSignal, addMapping, registerSignals, initMore, readAndRender
 - [x] Outcommented Code: Remove it or clearly state why outcommented.

Again thanks a lot. I did not look at it in action, however, I just came in contact with JS my self and therefore really have big respect for the difficult task you are tackling here. Kudos.

Thx,
@amstutz

klees commented:
Dear @nhaagen,

thanks for working on this, imo this already is a huge improvement. I know how hard this problem is and much appreciate your endurance.

Please implement the following changes:

- [ ] Lets use `withMainBarTreePosition(array $pos)` instead of [`withMainBarId`](https://github.com/ILIAS-eLearning/ILIAS/pull/2312/files#diff-5c7ddb1f728800a5ec75b9baeacf6d14R345). `$pos` should then contain the complete path down to the element in question, starting from the root node of the mainbar. Please also add some comment that explains what this is about. Please also rename [`enumerateNodes`](https://github.com/ILIAS-eLearning/ILIAS/pull/2312/files#diff-1bf7a0dfe0d0af761e9b8cd484e3a951R46) to `calculateMainBarTreePosition`. If you have doubts about the name lets discuss it. I'm no really satisfied with the name, but its better then the old one and I couldn't come up with an even better one.
- [ ] Please call `$f` [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2312/files#diff-5c7ddb1f728800a5ec75b9baeacf6d14R366). The method name would be lying otherwise. Pass ["tool", $counter] or ["entry", $counter] as first parameter to `$f` to distinguish between the two sections. This should play very well with the proposed change to `withMainBarId`.
- [ ] Please add a Roadmap-Entry that tells that `Tools` should get its own component in the KS. It most certainly is one...
- [ ] Please refactor [`renderMainbarEntry](https://github.com/ILIAS-eLearning/ILIAS/pull/2312/files#diff-1bf7a0dfe0d0af761e9b8cd484e3a951R60). The loop is almost incomprehensible and would benefit from [factoring out methods](https://github.com/ILIAS-eLearning/ILIAS/pull/2312/files#diff-1bf7a0dfe0d0af761e9b8cd484e3a951R95) and using [guard clauses](https://github.com/ILIAS-eLearning/ILIAS/pull/2312/files#diff-1bf7a0dfe0d0af761e9b8cd484e3a951R125). We could have a look together tomorrow and I'll show you in a little more detail which improvements I'm seeing.

For the JS-code we agreed to use a redux-like architecture, where events are applied to some datamodel and then the datamodel is projected onto the html. I definitely kind of see that, but I think we should improve the JS to show this architecture even more explicitely. This kind of goes in the same direction as @Amstutz remark about `jquery blood bath`, I guess. The JS still looks pretty ad hoc, and I would suggest the following improvements:

- [ ] Explain the desired architecture in some sentences on top of the modul.
- [ ] Add an internal structure via comments that has (approximately) these parts: `classes`, `state`, `reducer`, `rendering`, `wiring`, `persistence`. Distribute functions accordingly. I also suggest that we discuss that tomorrow.
- [ ] Follow @Amstutz  suggestions regarding `withAdditionalOnLoadCode`.

Thanks again and best regards!

matthiaskunkel commented:
Jour Fixe, 18 NOV 2019 : Richard notified us about an extension of a public interface. PR is accepted for ILIAS 6. Please merge.

nhaagen commented:
Thanks for your feedback and patience so far.
@Amstutz: I'd say this is good for another round.

Amstutz commented:
Hi @nhaagen,

thx for removing the todos. thx for cleaning up the js code in the renderers. thx for the improvement in naming to position.

Concerning the js: Darn it, I believe to have lost, the race for the cleanest JS interface in the UI Components up to this point. Nice job! The comments helped me. I know, Uncle Bob, Clean Code, heavily dislikes them, but here they seem to work nicely for me.

Things left: 
- [ ] Tests, they are failing.

@klees: My points are taken care of, except for the tests. Merge if you feel this is ready and the tests are fixed.

Thx a lot

nhaagen committed:
UI, Mainbar:  Revision

The goal is to have the entries of Mainbar, including subslates, adressable;
This opens the possibility to store engagement-states in a cookie.

The JS is way more structured and split up in model/renderer/persistence,
and the rendering is weeded out.
Slates are enumerated, and their signals are more aligned (note the public addOption-function in Signal).
